pkg/resource/state_machine/hooks.go

```diff
@@ -0,0 +1,174 @@
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	svcsdk "github.com/aws/aws-sdk-go/service/sfn"
+
+	svcapitypes "github.com/aws-controllers-k8s/sfn-controller/apis/v1alpha1"
+	commonutil "github.com/aws-controllers-k8s/sfn-controller/pkg/util"
+)
+
+// setResourceAdditionalFields queries and adds the tags to a StateMachine resource
+func (rm *resourceManager) setResourceAdditionalFields(
+	ctx context.Context,
+	ko *svcapitypes.StateMachine,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.setResourceAdditionalFields")
+	defer exit(err)
+
+	// Set StateMachine tags
+	ko.Spec.Tags, err = commonutil.GetResourceTags(
+		ctx,
+		rm.sdkapi,
+		rm.metrics,
+		string(*ko.Status.ACKResourceMetadata.ARN),
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// customUpdateStateMachine patches each of the resource properties in the backend AWS
+// service API and returns a new resource with updated fields.
+func (rm *resourceManager) customUpdateStateMachine(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (*resource, error) {
+	if delta.DifferentAt("Spec.Tags") {
+		err := commonutil.SyncResourceTags(
+			ctx,
+			rm.sdkapi,
+			rm.metrics,
+			string(*desired.ko.Status.ACKResourceMetadata.ARN),
+			latest.ko.Spec.Tags,
+			desired.ko.Spec.Tags,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if delta.DifferentExcept("Spec.Tags") {
+		err := rm.updateStateMachine(ctx, desired)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return desired, nil
+}
+
+func customPreCompare(
+	delta *ackcompare.Delta,
+	a *resource,
+	b *resource,
+) {
+	if len(a.ko.Spec.Tags) != len(b.ko.Spec.Tags) {
+		delta.Add("Spec.Tags", a.ko.Spec.Tags, b.ko.Spec.Tags)
+	} else if len(a.ko.Spec.Tags) > 0 {
+		if !commonutil.EqualTags(a.ko.Spec.Tags, b.ko.Spec.Tags) {
+			delta.Add("Spec.Tags", a.ko.Spec.Tags, b.ko.Spec.Tags)
+		}
+	}
+}
+
+// sdkUpdate patches the supplied resource in the backend AWS service API and
+// returns a new resource with updated fields.
+func (rm *resourceManager) updateStateMachine(
+	ctx context.Context,
+	desired *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.sdkUpdate")
+	defer func() {
+		exit(err)
+	}()
+	input, err := rm.newUpdateRequestPayload(ctx, desired)
+	if err != nil {
+		return err
+	}
+
+	_, err = rm.sdkapi.UpdateStateMachineWithContext(ctx, input)
+	rm.metrics.RecordAPICall("UPDATE", "UpdateStateMachine", err)
+	if err != nil {
+		return err
+	}
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := desired.ko.DeepCopy()
+
+	rm.setStatusDefaults(ko)
+	return nil
+}
+
+// newUpdateRequestPayload returns an SDK-specific struct for the HTTP request
+// payload of the Update API call for the resource
+func (rm *resourceManager) newUpdateRequestPayload(
```

Comment:
because this isn't generated, there is a danger of it becoming out of date when the controller is regenerated with new aws-sdk-go versions. Because of this, you might consider using a template in sdk_file_end custom hook point to output this function. Just FYI in case you want to do in a future PR.

Reply:
Oh yeah good idea :+1: